### ISSUE TITLE
Support of navigation properties

### ISSUE THREAD

**Issue by qtty1234** [bchavez/Bogus#115]
Here are example of classes:
```
public class Customer
{
    public int CustomerId { get; set; }
    public int Name { get; set; }
}
```

```
public class Order
{
    public int OrderId { get; set; }
    public int CustomerId { get; set; }
    public Customer Customer { get; set; }
}
```

And here is faker code.

```
var faker = new Faker<Order>()
    .RuleFor(x => x.Customer, () => new Faker<Customer>())
    .RuleFor(x => x.Customer.CustomerId, x => x.Random.Int(1, 1000))
    .Generate()
```

Is there an opportunity to support this case?

**Comment by bchavez** [comment]
Hi @qtty1234 ,

The way to do this is to first, fake out `Customer`, then reference the `Order` in a lambda to pull out the `Order.Customer.CustomerId`. Here's an example:

```
void Main()
{
   var fakeCustomers = new Faker<Customer>()
    .RuleFor(c => c.CustomerId, f => f.Random.Int(1, 1000))
    .RuleFor(c => c.Name, f => $"{f.Person.FirstName} {f.Person.LastName}");
   
   var fakeOrders = new Faker<Order>()
    .RuleFor(x => x.Customer, () => fakeCustomers )
    .RuleFor(x => x.CustomerId, (f, o) => o.Customer.CustomerId);
    
    fakeOrders.Generate().Dump();
}

public class Customer
{
   public int CustomerId { get; set; }
   public string Name { get; set; }
}

public class Order
{
   public int OrderId { get; set; }
   public int CustomerId { get; set; }
   public Customer Customer { get; set; }
}
```

![linqpad_925](https://user-images.githubusercontent.com/478118/34050438-bfdb96ca-e16f-11e7-9e17-54845957d7c9.png)

Hope hope that helps.

Thanks,
Brian
:zap: :runner: ***["I was struck by lighting. Walkin' down the street..."](https://www.youtube.com/watch?v=iypUpv9xelg)***

**Comment by qtty1234** [comment]
Hi @bchavez 

Thank you for reply.

Unfortunately it's not enough for my case. 
So, I'm going to explain what it would be nice to get.

I have a couple of rules wrapped into specific class (I've a little bit extended previous model):
```
public static class OrderFaker
{
  public static Faker<Order> Any()
  {
    var personFaker = new Faker<Person>()
      .RuleFor(x => x.PersonId, x => x.Random.Int(1, 100))
      .RuleFor(x => x.FullName, x => $"{x.Person.FirstName} {x.Person.LastName}");

    var somethingFaker = new Faker<Something>()
        .RuleFox(x => x.Id, x => x.Random(1, 100))
        .RuleFox(x => x.Value, x => x.Random.Word());

    var addressFaker = new Faker<Address>()
      .RuleFor(x => x.AddressId, x => x.Random(1, 100))
      .RuleFor(x => x.Something, () => somethingFaker);

    var customerFaker = new Faker<Customer>()
      .RuleFor(c => c.CustomerId, f => f.Random.Int(1, 1000))
      .RuleFor(c => c.Person, () => personFaker)
      .RuleFor(c => c.Address, () => addressFaker);

    var orderFaker = new Faker<Order>()
      .RuleFor(x => x.Customer, () => customerFaker)
      .RuleFor(x => x.CustomerId, (f, o) => o.Customer.CustomerId);

    return orderFaker;
  }
}
```

I'm going to use this code in two places and override some rules.

```
[Test]
public void Test1()
{
  // arrange
  var allowedValues = new[] {  "Word 1", "Word 2"  }; 
  var orderFaker = OrderFaker.Any()
    .RuleFor(x => x.Address.Something.Value, x => x.PickRandom(allowedValues));

  // act
  subject.Action(orderFaker.Generate());

  // assert
  // ...
}

[Test]
public void Test2()
{
  // arrange
  var allowedValues = new[] {  "Word 3", "Word 4"  }; 
  var orderFaker = OrderFaker.Any()
    .RuleFor(x => x.Address.Something.Value, x => x.PickRandom(allowedValues));

  // act
  subject.Action(orderFaker.Generate());

  // assert
  // ...
}
```

So, I see here two cases how to resolve my issue:

1) Get faker from `OrderFaker.Any()` and then modify rules explicitly.
```
// arrange
var allowedValues = new[] {  "Word 1", "Word 2"  }; 
var somethingFaker = new Faker<Something>()
  .RuleFor(x => x.Value, x => x.PickRandom(allowedValues));

var addressFaker = new Faker<Address>().RuleFox(x => x.Something, () => somethingFaker);
var orderFaker = OrderFaker.Any().RuleFox(x => x.Address, addressFaker);

// act
// etc...
```

2) Get faker, generate and than modify instance
```
// arrange
var allowedValues = new[] {  "Word 1", "Word 2"  }; 
var faker = new Faker();
var order = OrderFaker.Any().RuleFox(x => x.Address, addressFaker).Generate();
order.Address.Something.Value = faker.PickRandom(allowedValues);

// act
// etc...
```

You can notice that first case is a little bit redundant. The second one looks better but anyway we have to create extra `Faker` instance to perform `PickRandom(allowedValues)`.

So, it would be nice to have opportunity to set rules for nested fields. In this case arrange block will cleaner.

I hope my explanation make sense.

Thank,
Ivan.

**Comment by bchavez** [comment]
Hi Ivan,

Thank you for the clarification. I have a better understanding of your issue now.

Off the top of my head, I kinda feel `FinishWith` is the right approach here.

Ah, you're correct. It does seem `FinishWith` can only be called. Would allowing `FinishWith` to be called more than once (last call wins) work for you?

Thanks,
Brian

:walking_man: :walking_woman: **[Missing Persons - Walking in L.A.](https://www.youtube.com/watch?v=R_UpLtGEWoY)**

**Comment by qtty1234** [comment]
Hi @bchavez 

Thank you for your reply. The approach with multiple `FinishWith` works for me. 
I would suggest to add overload for FinishWith with selector and faker.

From my point of view this looks better
```
var allowedValues = new[] {  "Word 1", "Word 2"  }; 
OrderFaker.Any()
  .FinishWith(x => Address.Something.Value, f => f.PickRandom(allowedValues))
```

than this one
```
var allowedValues = new[] {  "Word 1", "Word 2"  }; 
OrderFaker.Any()
  .FinishWith( (x, f) => x.Address.Something.Value = f.PickRandom(allowedValues);)
```

Thank you for assistance
Ivan

**Comment by bchavez** [comment]
Hi Ivan,

Just published **Bogus** `v21.0.5` with recent changes to allow calling `FinishWith` multiple times. https://www.nuget.org/packages/Bogus/21.0.5

For now, the second `.FinishWith( (x, f) =>...)` is what we'll use. I know it's not cleanest for you, but hopefully, `v21.0.5` should unblock what you are trying to achieve with **Bogus**. Let me know how it goes.

As for the first solution, `.FinishWith(x => x.NestedObj.Prop, f => ..)` I need more time to think about the overall consequences, changes, complexity, semantics, and code maintenance for allowing nested/chained member accessor rules.

Again, thank you for bringing your issue to my attention.

Thanks,
Brian

:walking: :sunglasses: ***["So don't delay... act now supplies are running out..."](https://www.youtube.com/watch?v=LQj--Kjn0z8)***

**Comment by taiwoa** [comment]
@bchavez What if Order is defined with a Customer list; ?

```
public class Order
{
   public int OrderId { get; set; }
   public int CustomerId { get; set; }
   public Customer[] Customers { get; set; }
}
```

How would RuleFor Order.Customers be declared?

**Comment by skaie7** [comment]
I have managed to work around this issue by setting nested properties in `Rules()` instead of `RuleFor()`, for example:

```C#
var faker = new Faker<Order>()
    .Rules((o, f) => o.Customer.CustomerId = f.Random.Int(1, 1000));
```

In my case it's impossible to create a separate Faker for the `Customer` field as it is a read-only property

**Comment by Redplcs** [comment]
@xzxzxc I definitely will support you in your PR, because I've faced this issue on my project. Why the hell I can't just use `RuleFor` method for nested types just to be it consistent. This bellow looks way simpler visually, than instantiating other fakes just to fake nested type

```csharp
var faker = new Faker<Person>()
    .RuleFor(p => p.FirstName, f => f.Name.FirstName())
    .RuleFor(p => p.Job.Address, f => f.Address.FullAddress());
```
